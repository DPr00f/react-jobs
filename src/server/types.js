/* @flow */

export type ReactJobsClientContext = {
  nextJobID: () => number,
  popJobRehydrationForSRR: (jobID: number) => any,
};

export type JobState = {
  inProgress: boolean,
  result?: any,
  error?: any,
}

export type RehydrateState = {
  jobsState: { [key: number] : JobState },
};

export type ReactJobsServerContext = {
  nextJobID: () => number,
  registerJobState: (number, JobState) => void,
  getJobState: number => any,
};

export type RenderContext = {
  registerJobState: (number, JobState) => void,
  getState: () => RehydrateState,
};

export type ClientProviderContext = {
  reactJobsClient: ReactJobsClientContext,
};

export type ServerProviderContext = {
  reactJobsServer: ReactJobsServerContext,
};

export type ProviderContext = ServerProviderContext & ClientProviderContext;
