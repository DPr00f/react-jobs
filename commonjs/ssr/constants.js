'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
/* eslint-disable import/prefer-default-export */

var STATE_IDENTIFIER = exports.STATE_IDENTIFIER = '__REACT_JOBS_STATE__';