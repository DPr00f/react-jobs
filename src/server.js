/* @flow */

import runJobs from './runJobs';

export default {
  runJobs,
};
