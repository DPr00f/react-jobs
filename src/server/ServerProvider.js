/* @flow */

import { Children, Component, Element, PropTypes } from 'react';
import type { ServerProviderContext, RenderContext } from './types';

type React$Element = Element<*>;

type Props = {
  children?: React$Element,
  renderContext: RenderContext,
};

let currentjobID = 0;
const jobResults = {};

class ServerProvider extends Component {
  props: Props;

  getChildContext() {
    const context : ServerProviderContext = {
      reactJobsServer: {
        nextJobID: () => {
          currentjobID += 1;

          if (this.props.renderContext) {
            this.props.renderContext.setCurrentJobID(currentjobID);
          }

          return currentjobID;
        },
        registerJobResults: (jobID: number, results: any) => {
          jobResults[jobID] = results;
          if (this.props.renderContext) {
            this.props.renderContext.registerResolvedJob(jobID);
          }
        },
        getJobResults: (jobID: number) => jobResults[jobID],
      },
    };

    return context;
  }

  render() {
    return Children.only(this.props.children);
  }
}

ServerProvider.childContextTypes = {
  reactJobs: PropTypes.object.isRequired,
};

export default ServerProvider;
