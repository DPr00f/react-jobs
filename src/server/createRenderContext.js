/* @flow */

import type { RenderContext } from '../sharedTypes';

export default function createRenderContext() : RenderContext {
  let currentjobID = 0;
  const resolvedJobs = {};
  return {
    setCurrentJobID: (jobID) => {
      currentjobID = jobID;
    },
    registerResolvedJob: (jobID) => {
      resolvedJobs[jobID] = true;
    },
    getState() {
      return {
        currentjobID,
        resolvedJobs,
      };
    },
  };
}
