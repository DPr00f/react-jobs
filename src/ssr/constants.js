/* eslint-disable import/prefer-default-export */

export const STATE_IDENTIFIER = '__REACT_JOBS_STATE__'
