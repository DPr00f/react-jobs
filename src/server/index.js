/* @flow */

import createRenderContext from './createRenderContext';
import runJobs from './runJobs';
import ServerProvider from './ServerProvider';

export default {
  createRenderContext,
  runJobs,
  ServerProvider,
};
