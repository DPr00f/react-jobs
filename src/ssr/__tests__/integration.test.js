/* @flow */

import React from 'react';
import { renderToString } from 'react-dom/server';
import { Foo, resolveAfter, rejectAfter } from '../../../tools/tests/helpers';
import {
  withJob,
  runJobs,
  rehydrateJobs,
} from '../';
import { STATE_IDENTIFIER } from '../constants';

const workTime = 10;

const createComponents = () => ({
  Hello: withJob(() => resolveAfter(workTime, 'Hello world!'))(Foo),
  Goodbye: withJob(() => resolveAfter(workTime, 'Goodbye world!'))(Foo),
  Fail: withJob(() => rejectAfter(workTime, 'Oh noes!'))(Foo),
});

const createApp = () => {
  const { Hello, Goodbye, Fail } = createComponents();
  return (
    <div>
      <Hello />
      <Goodbye />
      <Fail />
    </div>
  );
};

describe('integration', () => {
  afterEach(() => {
    delete global.window[STATE_IDENTIFIER];
  });

  it('full ssr app works', () => {
    const serverApp = createApp();
    return runJobs(serverApp)
      .then(({ appWithJobs, state, STATE_IDENTIFIER: STATE_ID }) => {
        // "Server" render
        const serverRender = renderToString(appWithJobs);
        expect(serverRender).toMatchSnapshot();
        // Attach the state to the "window" for the client
        global.window[STATE_ID] = state;
        return serverRender;
      })
      .then((serverRender) => {
        // "Client" render
        const clientApp = createApp();
        return rehydrateJobs(clientApp)
          .then(({ appWithJobs }) => {
            const clientRender = renderToString(appWithJobs);
            expect(clientRender).toEqual(serverRender);
            return appWithJobs;
          });
      })
      .then((clientApp) => {
        // Second "Client" render
        // The second "Client" render should be async as we only use the
        // server hydration state once.
        const clientSecondRender = renderToString(clientApp);
        expect(clientSecondRender).toMatchSnapshot();
      });
  });
});
