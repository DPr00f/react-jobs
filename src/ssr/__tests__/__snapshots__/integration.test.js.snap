// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration full ssr app works 1`] = `"<div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"-513584873\\"><div data-reactid=\\"2\\">Hello world!</div><div data-reactid=\\"3\\">Goodbye world!</div><div data-reactid=\\"4\\"><!-- react-text: 5 -->Error: <!-- /react-text --><!-- react-text: 6 -->Oh noes!<!-- /react-text --></div></div>"`;

exports[`integration full ssr app works 2`] = `"<div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"-2088486596\\"><div data-reactid=\\"2\\">In progress...</div><div data-reactid=\\"3\\">In progress...</div><div data-reactid=\\"4\\">In progress...</div></div>"`;
